typespec-extension/src/code-model-builder.ts

```diff
@@ -1987,10 +1988,30 @@ export class CodeModelBuilder {
     const discriminator = getDiscriminator(this.program, type);
     if (discriminator) {
       discriminatorPropertyName = discriminator.propertyName;
+      // find the discriminator property from model
+      // the property is required for getting its serializedName
+      let discriminatorProperty = Array.from(type.properties.values()).find(
+        (it) => it.name === discriminatorPropertyName,
+      );
+      if (!discriminatorProperty) {
+        // try find the discriminator property from any of its derived models
+        for (const deriveModel of type.derivedModels) {
+          discriminatorProperty = Array.from(deriveModel.properties.values()).find(
+            (it) => it.name === discriminatorPropertyName,
+          );
+          if (discriminatorProperty) {
+            break;
+          }
+        }
+      }
+      const discriminatorPropertySerializedName = discriminatorProperty
+        ? this.getSerializedName(discriminatorProperty)
```

Comment:
`this.getSerializedName(discriminatorProperty)` would give the serialized name (it differs in the case of `projectedName("json",)`)